File: exam_at_home/2/euler_tour.cc
/*
* author: wang-yang
* email: tnst92002@gmail.com
*/
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> V;
void dfs(int v) {
    cout << ' ' << v;
    for (int u:V[v])
        dfs(u);
    cout << ' ' << v;
}

int main() {
    int N;
    cin >> N;
    V.resize(N);

    for (int i=0;i<N;++i) {
        int v, u;
        cin >> v;
        while (v--) {
            cin >> u;
            V[i].emplace_back(u);
        }
    }

    cout << "Euler tour:";
    dfs(0);
    cout << '\n';
}

File: exam_at_home/2/euler_tour.py
# author: jamie
# email: jinjiedeng.jjd@gmail.com
def EulerTour (adj, euler_tour):
    stack = [(0, 0)]
    while len(stack):
        u, i = stack.pop()
        if i != len(adj[u]):
            stack.append((u, i + 1))
            stack.append((adj[u][i], 0))
        if i == 0:
            euler_tour.append(u)
        if i == len(adj[u]):
            euler_tour.append(u)

if __name__ == "__main__":
    n = int(input())
    adj = []
    for _ in range(n):
        _, *children = map(int, input().split())
        adj.append(children)
    euler_tour = []
    EulerTour(adj, euler_tour)
    print("Euler tour:", " ".join(map(str, euler_tour)))